URL service docs (#125697)
elastic/kibana: src/plugins/share/README.mdx

---
id: kibDevSharePluginReadme
slug: /kibana-dev-docs/tutorials/share
title: Kibana share Plugin
summary: Introduction to Kibana "share" plugin services (sharing popup menu, URL locators, and short URLs).
date: 2022-02-21
tags: ['kibana', 'onboarding', 'dev', 'tutorials', 'share', 'url', 'short-url', 'locator']
---

# Share plugin

The `share` plugin contains various utilities for displaying sharing context menu,
generating deep links to other apps using *locators*, and creating short URLs.

## Sharing context menu

You can register an item into sharing context menu (which is displayed in
Dashboard, Discover, and Visualize apps).

### Example registration

```ts
import { ShareContext, ShareMenuItem } from 'src/plugins/share/public';

plugins.share.register({
  id: 'demo',
  getShareMenuItems: (context) => [
    {
      panel: {
        id: 'demo',
        title: 'Panel title',
        content: 'Panel content',
      },
      shareMenuItem: {
        name: 'Demo list item (from share_example plugin)',
      },
    }
  ],
});
```

Now the "Demo list item" will appear under the "Share" menu in Discover and

> Dashboard applications.

Reviewer: worth mentioning how to get it to display in other applications ? (how to add that menu to your app)

## Locators

*Locators* are entities which given some *parameters* can navigate Kibana user
to some deep link in Kibana or return back a full formatted URL.

Kibana apps create locators and expose them from their plugin contracts. Then
other plugins can use those locators to navigate deeply into Kibana applications.

> Locators work, both, on the server and browser. However, it is the responsibility

Reviewer 2: ```suggestion
Locators work on the server and browser. However, it is the responsibility
```

> of the app, which provides the locator, to register and expose the same locator

Reviewer: worth putting this into a "warning" box to make it stand out, i think its very important that you register locators on both server and client and that one locator with its id is the same on both sides

on the server and browser.

### Creating a locator for your Kibana app

A Kibana app can create one or more *locators*, which given an object of *parameters*
knows how to specify return a *Kibana location* 3-tuple. The 3-tuple consists of:
(1) the app ID; (2) a path within the app; (3) a serializable *location state* object.
The Kibana location 3-tuple is used to navigate the Kibana app in SPA (single page app)
way&mdash;without reloading the page.

```ts
import type { SerializableRecord } from '@kbn/utility-types';
import { LocatorDefinition, LocatorPublic } from 'src/plugins/share/public';

export interface HelloLocatorParams extends SerializableRecord {
  name: string;
}

export type HelloLocator = LocatorPublic<HelloLocatorParams>;

export class HelloLocatorDefinition implements LocatorDefinition<HelloLocatorParams> {
  public readonly id = 'HELLO_LOCATOR';

  public readonly getLocation = async ({ name }: HelloLocatorParams) => {
    return {
      app: 'sampleAppId',
      path: `/hello?name=${encodeURIComponent(name)}`,

>       state: {},

Reviewer: i think it would be worth putting some actual state here, we prefer iusing this state to url encoding things into `path`

    };
  };
}
```

Once your locator has been defined, you need to register it in the central locator
registry with the `share` plugin:

```ts

> const locator = plugins.share.url.locators.create(new HelloLocatorDefinition());

Reviewer: worth mentioning why you need to register (instead of only exposing it on the contract)

```

> You can then return the `locator` reference from your plugin contract fro consumption

Reviewer 2: ```suggestion
You can then return the `locator` reference from your plugin contract for consumption
```

> by other plugins:

Reviewer: should it be part of setup or start contract ?

```ts
class MyPlugin {
  setup(core, plugins) {
    const locator = plugins.share.url.locators.create(new HelloLocatorDefinition());

    return {
      locator,
    };
  }
}
```

### Using locator of another app

First you need to get access of the locator you want to use. Best way to do it is

> by acquiring it from the plugin contract of the app you are interested to navigate

Reviewer 2: ```suggestion
by acquiring it from the plugin contract of the app you want to navigate
```

to. For example, you can get hold of the Dashboard app locator from the `dashboard`
plugin contract:

```ts
class MyPlugin {
  setup(core, plugins) {
    const dashboardLocator = plugins.dashboard.locator;
  }
}
```

Alternatively, you can also acquire any locator from the central registry in
the `share` plugin using the ID of the locator. This is useful, if you run into

> circular dependency issues, however, it is better to acquire the locator directly

Reviewer: i am not sure i would mention this, we don't want ppl to use locators from the registry imo unless they have a usecase where they don't know the locator they'll be using at compile time.

from the plugin contract as above, which makes dependencies explicit.

You can use the `.get()` method on the locators service to acquire the locator
from the central registry:

```ts
class MyPlugin {
  setup(core, plugins) {
    const dashboardLocator = plugins.share.url.locators.get('DASHBOARD_APP_LOCATOR');
  }
}
```

> Once have have a reference to the desired locator object, you can use that locator

Reviewer 2: ```suggestion
Once you have a reference to the desired locator object, you can use that locator
```

to navigate to various pages in Kibana or use it to build a URL string.

The best way to use the locator is to use its `.navigateSync()` method to navigate

> in SPA way (without page reload):

Reviewer 2: Maybe we can explain why this is the best way?

```
The `.navigateSync()` method is the best way to navigate in SPA way (avoiding a page reload). <It blah blah>:
```

Or if it is the only way
```
Use the `.navigateSync()` method to navigate in an SPA way (avoiding a page reload):
```

```ts
dashboardLocator.navigateSync({ dashboardId: '123' });
```

The above will immediately navigate to the Dashboard app, to the dashboard with
ID `123`.

Another useful feature of locators is their ability to format a URL string of the
destination. The best way to do it is using the `.getRedirectUrl()` method
of locators.

```ts
const url = dashboardLocator.getRedirectUrl({ dashboardId: '123' });
```

The above will generate a *redirect* URL, which also preserves the serializable
*location state* object, which will be passed to the destination app.

Alternatively, for compatibility with old URL generators, you can also use the
`.getUrl()` method of locators to generate just the final app destination URL
string without the location state object.

```ts
const url = await dashboardLocator.getUrl({ dashboardId: '123' });
```

> Note, in this process you lose the location state object, which effectively

Reviewer: put this in a "warning" block as well, its worth highlighting this is a legacy thing and should not be used in the new code.

makes the final destination incorrect, if the app makes use of the location
state object.

For more ways working with locators see the `LocatorPublic` interface.

> 

Reviewer: another thing worth mentioning is how to handle migrations and references

> 

Reviewer: i think we are missing a part where we explain how the application which registered locator should handle the incoming state

## Short URLs

The `share` plugin has Short URL service, which allows to create Kibana short
URLs. The service with the same interface is available, both, on the server
and browser.

When creating a short URL it is possible to use a custom short *slug*, used
to dereference the short URL. Also, a human-readable short slug can be
automatically generated.

First, to be able to create a short URL, one needs to get hold of a locator
reference, for which the short URL will be created. (See above section on
*Locators* for more info.) Lets say we want to create a short URL for the
Dashboard application, for that, we first acquire reference to the dashboard
locator:

```ts
class MyPlugin {
  setup(core, plugins) {
    const dashboardLocator = plugins.dashboard.locator;
  }
}
```

Now you can create a short URL using your locator and the short URL service:

```ts
const shortUrls = plugins.share.url.shortUrls.get(null);
const url = await shortUrls.create({
  locator,
  params: {
    dashboardId: '123',
  }
});
```

You can make the short URL slug human-readable by specifying the

> `humanReadableSlug` flag:

Reviewer 2: silly question: what happens when both `slug: "cool-slug"` and `humanReadableSlug: true` are provided?
